src/selectors/entities/channel_categories.ts

```diff
@@ -189,6 +189,98 @@ export function makeFilterAutoclosedDMs(getAutocloseCutoff = getDefaultAutoclose
     );
 }
 
+export function makeFilterAutoclosedDMs(): (state: GlobalState, channels: Channel[], categoryType: string) => Channel[] {
+    return createSelector(
+        (state: GlobalState, channels: Channel[]) => channels,
+        (state: GlobalState, channels: Channel[], categoryType: string) => categoryType,
+        getCurrentChannelId,
+        (state: GlobalState) => state.entities.users.profiles,
+        getCurrentUserId,
+        getMyChannelMemberships,
+        (state: GlobalState) => getInt(state, Preferences.CATEGORY_SIDEBAR_SETTINGS, Preferences.LIMIT_VISIBLE_DMS_GMS),
+        (originalChannels, categoryType, currentChannelId, profiles, currentUserId, myMembers, limitPref) => {
+            if (categoryType !== CategoryTypes.DIRECT_MESSAGES) {
+                // Only autoclose DMs that haven't been assigned to a category
+                return originalChannels;
+            }
+
+            const channels = originalChannels.slice();
+
+            function sortByLastViewedAt(channelA: Channel, channelB: Channel) {
+                const channelALastViewedAt = myMembers[channelA.id]?.last_viewed_at || 0;
+                const channelBLastViewedAt = myMembers[channelB.id]?.last_viewed_at || 0;
+
+                if (channelBLastViewedAt === channelALastViewedAt) {
+                    return 0;
+                }
+
+                if (channelALastViewedAt === 0) {
+                    return -1;
+                }
+
+                if (channelBLastViewedAt === 0) {
+                    return 1;
+                }
+
+                return Math.sign(channelBLastViewedAt - channelALastViewedAt);
```

Comment:
Strictly speaking, I don't think we need `Math.sign` here since any positive or negative values are treated as 1 and -1 respectively. I do have a funny story about a production bug I caused with this sort of code at my previous job though...